worker/worker.py

```diff
@@ -270,6 +277,130 @@ def get_credentials(config, options, args):
     return username, password
 
 
+def verify_required_cutechess(testing_dir, cutechess):
+    cutechess = os.path.join(testing_dir, cutechess)
+
+    print("Obtaining version info for {} ...".format(cutechess))
+
+    if not os.path.exists(cutechess):
+        print("{} does not exist ...".format(cutechess))
+        return False
+
+    try:
+        with subprocess.Popen(
+            [cutechess, "--version"],
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            universal_newlines=True,
+            bufsize=1,
+            close_fds=not IS_WINDOWS,
+        ) as p:
+            errors = p.stderr.read()
+            pattern = re.compile("cutechess-cli ([0-9]*).([0-9]*).([0-9]*)")
+            major, minor, patch = 0, 0, 0
+            for line in iter(p.stdout.readline, ""):
+                m = pattern.search(line)
+                if m:
+                    print("Found: ", line.strip())
+                    major = int(m.group(1))
+                    minor = int(m.group(2))
+                    patch = int(m.group(3))
+    except (OSError, subprocess.SubprocessError) as e:
+        print("Unable to run cutechess-cli. Error: {}".format(str(e)))
+        return False
+
+    if p.returncode != 0:
+        print(
+            "Unable to run cutechess-cli. Return code: {}. Error: {}".format(
+                format_return_code(p.returncode), errors
+            )
+        )
+        return False
+
+    if major + minor + patch == 0:
+        print("Unable to find the version of cutechess-cli.")
+        return False
+
+    if (major, minor) < (1, 2):
+        print("Requires cutechess 1.2 or higher, found version doesn't match")
+        return False
+
+    return True
+
+
+def setup_cutechess():
+    # Create the testing directory if missing.
+    worker_dir = os.path.dirname(os.path.realpath(__file__))
+    testing_dir = os.path.join(worker_dir, "testing")
+    if not os.path.exists(testing_dir):
+        os.makedirs(testing_dir)
+
+    try:
+        os.chdir(testing_dir)
+    except Exception as e:
+        print("Unable to enter {}. Error: {}".format(testing_dir, str(e)))
+        return False
+
+    cutechess = "cutechess-cli" + EXE_SUFFIX
+
+    if not verify_required_cutechess(testing_dir, cutechess):
+        if len(EXE_SUFFIX) > 0:
+            zipball = "cutechess-cli-win.zip"
+        elif IS_MACOS:
+            zipball = "cutechess-cli-macos-64bit.zip"
+        else:
+            zipball = "cutechess-cli-linux-{}.zip".format(platform.architecture()[0])
+        try:
+            download_from_github(zipball, testing_dir)
+            zip_file = ZipFile(zipball)
```

Comment:
Context manager added in python 3.2, see https://docs.python.org/3.5/library/zipfile.html#zipfile.ZipFile

```python
            zip_file = ZipFile(zipball)
            zip_file.extractall()
            zip_file.close()
```

```python
            with ZipFile(zipball) as zip_file:
                zip_file.extractall()
```

Reply:
Changed!